Cluster High-availability Mode API

dhellmann opened the pull request:
As we add more deployment topologies for OpenShift, we need a way to
communicate expectations to the operators managing the cluster. This
enhancement describes an API for telling operators how hard they
should work to provide high-availability.

dhellmann commented:
/cc @deads2k @wking @eparis @hexfusion @mrunalp @romfreiman

dhellmann commented:
/cc @eranco74

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the cluster capabilities (at
+  least for now).
```
It might be worth a ```go section in the proposal with the content you expect to add to openshift/api for this.  That would make it clear that your new properties were going to appear in a less-mutable `status` fields vs. user-configurable `spec` fields.  Precedent for using `status` properties for write-once fields in the Infrastructure object.

dhellmann replied:
@wking I'm not sure what you mean here. Are you asking for a go struct definition?

dhellmann replied:
I think the immutability is covered with the prose now. Let me know if it's still not clear and we can do more.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+operation. Systems such as the assisted installer, which may
+experience periods early in the lifetime of the cluster when full HA
+is not possible, can start with `none` and transition to `full` when
+the cluster is more fully online.
```
This conflicts with the "Provide a mutable API for changing the cluster capabilities..." non-goal.  I think it's fine to drop that as a non-goal, and to have HA-ness in a `spec` property.  Then... something... maybe [the config operator][1] would validate incoming `spec` properties and pass valid choices through to `status` properties.  Other components would consume the `status` components (so the validation buffers them from a mass explosion if a user typos and edit to a `spec` property).  Once you have that sort of dynamic control, you probably want `status.conditions[]` as well, so the config operator has space to file complaints with details about any `spec` values that it is mad about.

[1]: https://github.com/openshift/cluster-config-operator

cgwalters replied:
Hmm yeah, supporting dynamic transitions is a whole other thing.  Is there a problem today with the assisted installer having operators in degraded mode until everything comes online?

An immutable property seems better to me.

eranco74 replied:
Currently, assisted-installer is using operator-specific APIs to tell the CEO to start the etcd cluster without waiting for 3 master nodes (the bootstrap will pivot to be the 3rd master node once the control plane is running on the first 2 masters).
The assisted-installer has a similar issue with kubernetes-OVN (which won't configure OVS without 3 master nodes (it actually looks at the number control plane replica in cluster-config configMap).
These components functionality is mandatory for the control plane to start on the 2 master nodes and for the installation to complete.

dhellmann replied:
I dropped the discussion of special handling for the assisted installer workflow and tried to emphasize more the immutable nature of the first version of the API.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+is not possible, can start with `none` and transition to `full` when
+the cluster is more fully online.
+
+The `resourceLimits` fields will use an enum to express the general
```
So are we not breaking this down by resource type ([`cpu`, `memory`, etc.][1])?  On the one hand, it's not clear to me how many components can scale their consumption of different resource types independently (e.g. reducing memory use while leaving CPU consumption high).  On the other hand, it would be a shame to scale down a CPU-intensive but memory/disk-light component if you had plenty of CPU but limited amounts of memory.  It may be worth going through the heavy components (monitoring, etc.) and providing more details about their current consumption and the expected reductions needed to fit into some hypothetical target system.

[1]: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/

dhellmann replied:
Yeah, we did talk about using separate fields for different resource types, but it seemed better to keep it simple.

cgwalters replied:
I like the intent of this but it feels like it's going to be subject to a lot more debate - maybe this one is a separate enhancement?  I'd like to get the centralized field for single node first personally.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
```
this section still needs filling in.  I think the main risk is that we flub the initial schema and end up providing an API that no components actually need.  Might be worth compiling a list of the existing config properties we're replacing (just the etcd and auth HA flags?) and the per-component configs we are avoiding (some monitoring knob?  Anything else?).  The conservative approach would be to leave these knobs per-component until we reach some critical mass to make a stronger "look at all these per-component knobs we can quietly de-emphasize by growing this cluster-scoped knob" case.

dhellmann replied:
The driver for this requirement is the need to reduce the overall footprint of the cluster, rather than any specific component, right?

How would the installer set the component-specific API fields?

dhellmann replied:
I've added details, but I honestly don't see a way to do this with operator-specific APIs that doesn't involve a huge amount of work to tie the installer to those APIs so they can be set during deployment.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this
```
This seems tricky.  What happens if we grow a new HA enum (say), and the user updates to it before all cluster components understand that value?  Looks like the config operator is currently [at level 30][1], so before the bulk of operator at level 50, but after etcd ([at level 12][2]) and other early operators.  That means that a config operator would learn of the new HA enum early, and could accept the user setting it before the update completed and later operators understood what to do about the new enum.  If we never update in any of the non-HA modes, that's ok.  But the same issues would apply to `resourceLimits` if that's a `spec`-configurable knob.  Maybe components could scrape `Capabilities`'s `status` and reflect their currently-accepted value in their own `status` (on their ClusterOperator or a per-component config object or whatever).  That way they could keep rolling forward with whatever their previously accepted config was if the central value was unrecognized, hoping that soon they would be rebooted into a new version which understood the newly-accepted central value.

[1]: https://github.com/openshift/cluster-config-operator/tree/351f7aa07c0886866484f9916ebb2b2c5663c0e1/manifests
[2]: https://github.com/openshift/cluster-etcd-operator/tree/06667d94d6313d6edd59801a351de7038987f36d/manifests

dhellmann replied:
Maybe that's another argument for making the API immutable, at least to start?

dhellmann replied:
Thinking about this more, it seems error prone to put logic to manage this in every operator that consumes the API. I've captured that idea in the alternatives, and proposed a way to manage the settings in the new API to avoid the duplication of effort.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+We could add the fields we know about today to the
+`infrastructure.config.openshift.io` API, instead of creating a new
+API. This would be less work, initially, but muddies the meaning of
+the APIs.
```
Current `Infrastructure` scope [is][1]:

> Infrastructure holds cluster-wide information about Infrastructure.

Which is already fairly muddy ;).  It's not clear to me that HA-ness or target resource consumption is less infrastructure-y than [API URIs][2], and those are already part of the `Infrastructure` spec.  I think the trade-off here is between recycling an existing CRD (less dev and doc work, possibly slight efficiency savings in API-server memory or something?  I don't have a clear understanding of per-CRD resource use) and creating a new one (ideally clearer scoping, if we're worried about overloading an existing CRD).

[1]: https://github.com/openshift/api/blob/c6debb38648fa2094d36ef7dd2ffc0dca97da879/config/v1/types_infrastructure.go#L10
[2]: https://github.com/openshift/api/blob/c6debb38648fa2094d36ef7dd2ffc0dca97da879/config/v1/types_infrastructure.go#L70-L79

dhellmann replied:
Yeah, I don't have a particularly strong opinion on this. It would be less work in the short term to just extend the existing resource, but it seemed cleaner to add a new one. I'll go along with whatever the API owners agree to do.

wking commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+since each component could be configured separately, but would make
+integrating with the installer and cluster profiles much more
+complicated since each component would need to be configured
+separately.
```
Do we have a realistic number of component-specific knobs?  So far I've heard about etcd and auth knobs for HA-ness, and monitoring knobs for resource consumption.  Three knobs is not "much more complicated" than a central knob in my subjective opinion.  It's certainly possible that there are other component-specific knobs out there today, or planned for the future if we don't have these cluster-scoped knobs, but I think it's worth digging into that a bit more to make a stronger case for cluster-scoped knobs.  And as you point out with the precedence comments in the *Proposal* section, having cluster-scoped knobs still leaves space for component-scoped knobs, for components that want independent flexibility.

dhellmann replied:
The implementation details section of #504 mentions cluster-ingress-operator and machine-config-operator as needing to honor the API. Neither has their own API for these behaviors today.

dhellmann replied:
I added the operators we know will be affected to this section of the doc.

dhellmann commented:
/cc @lilic

dhellmann commented:
/cc @cgwalters

cgwalters reviewed (approved):
Thanks so much for writing this!

romfreiman commented:
@danwinship @dcbw @alexanderConstantinescu we should explore using this api for that assisted installer flow in 4.8 (2 nodes that become 3 later on).

lilic commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,328 @@
+`large`. The default is `large`, representing the behavior for
+operators today. The `medium` configuration would be used to lower the
+resource use in single-node production clusters. For example, the
+cluster-monitoring-operator might deploy prometheus with a reduced set
```
This makes sense, but determining what we do not need, e.g. what is reduced might be harder to achieve. This also depends on other components in the cluster that expose metrics we scrape, so something like setting sample limits to each target/component might be a good start.

I think we would need more specific resource requirements to know what medium and small means here, in this KEP.

dhellmann replied:
It does seem to be the consensus that we should move the sizing discussion to another enhancement, so I've cut this part of the text out for now.

lilic reviewed (commented):
cc @simonpasquier @bbrowning for 👀 on the monitoring side of things

dhellmann commented:
@sttts This enhancement is related to the Jira ticket we discussed earlier.

dhellmann commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,336 @@
+`capabilities` API to change settings that use new values not
+supported in the older version.
+
+### Version Skew Strategy
```
@mrunalp @staebler I added some details about handling version skew based on our conversation earlier today. Let me know if I've captured it accurately here.

mrunalp replied:
@dhellmann Looks good. Thanks!

smarterclayton commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,337 @@
+
+A new [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+`capabilities.config.openshift.io`, is created to hold details about
```
I don't like the idea of a bunch of different capabilities being in one object.  I would expect many characteristics to be in infrastructure status.  HA mode I expect on infrastucture, for instance.

derekwaynecarr replied:
by putting it in status its also fixed immutable, which i like.

we also need to have graceful fallback when not present.

smarterclayton commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,337 @@
+start, and may be expanded over time.
+
+The `highAvailabilityMode` field will use an enum to express the
+high-availability expectations. Possible values are `full` and
```
nit: All constants have to follow upstream constants conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#constants

smarterclayton commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,337 @@
+should not configure the operand for highly-available
+operation.
+
+The new cluster-wide `capabilities` API will supersede existing
```
If we only have one capability, and it can fit in infrastructure, I don't see a reason to have the capabilities object yet.

derekwaynecarr replied:
agree.

dhellmann replied:
We have identified 2 other candidates for capabilities.

One is a flag to indicate whether operators should deploy their GUI components. That could be part of an API that the console-operator defines, but the monitoring operator would need to use it and the installer would need to set it, so putting it in the capabilities seemed to make sense.

The other is some expression of the "size" operators should assume for their operands, so they could be reconfigured for smaller footprint deployments. We have not yet agreed on how to express that, and didn't want to hold this work up while we figured it out, so it was removed from this PR.

If those also seem like "infrastructure" settings, then we can rewrite this PR. But if we're going to have capabilities at some point, I would just as soon do it from the start.

smarterclayton commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,337 @@
+
+We could add the fields we know about today to the
+`infrastructure.config.openshift.io` API, instead of creating a new
+API. This would be less work, initially, but muddies the meaning of
```
I don't agree that it muddies the meaning of the API.  It certainly fits within the broad strokes of the existing object, and so from a consistency perspective if "the control plane's HA property" relates to the infrastructure needed for the cluster, it belongs on infrastructure (we are parsimonious about objects, and we are skeptical in general of generic objects with a grab bag of generic fields).

derekwaynecarr replied:
with externalized control plane use cases, its also fine that we project the Infrastructure CR in the target cluster, so I kind of prefer not having a new object.

/cc @csrwng

dhellmann commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,357 @@
+operator-specific values are not set, a replacement for unsupported
+APIs, and a way to avoid adding new operator-specific APIs.
+
+### Operator Lifecycle Manager Operators
```
@mrunalp @ecordell Does this section match what you had in mind for the OLM? We don't have to work out all of the details, as long as we have a general approach in mind.

derekwaynecarr commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,357 @@
+API. This would be less work, initially, but muddies the meaning of
+the APIs.
+
+### Expose the cluster profile name directly
```
would love if after all is said and done, we end up with 2 and are able to collapse the stragglers with this signal.

derekwaynecarr commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,357 @@
+would be less discoverable and we would not be able to offer users a
+way to change them at runtime (if we ever want to do that).
+
+### Recognize HA/non-HA configuration based on control plane node count
```
this is also confusing with externalized control plane, where you have no control plane hosts in the cluster, and a variable number of workers to meet the workload demand.  prefer projecting via the Infrastructure CR.
//cc @csrwng

dhellmann replied:
Good point, I'll add this to the doc.

zanetworker commented on enhancements/cluster-capabilities-api.md:
```diff
@@ -0,0 +1,357 @@
+
+### Operator Lifecycle Manager Operators
+
+Operators managed by OLM should not be required to respond to an
+OpenShift-specific API. Therefore, the OLM should read the API and add
+an environment variable with the high-availability mode, similar to
+how the proxy settings are configured today.
```
Reading this, it is not clear to me what happens next.

What I understood so far is that there will be an Env variable, my question is, how is this ENV read, can we elaborate by giving an example?

dhellmann replied:
The operator managed by OLM would read the environment variable, if it decides to honor it. I was told this was similar to some other settings that OLM already passes to the operators it manages.

zanetworker replied:
Does this mean that modification in the operator logic is required? IIUC in some cases (e.g., HA) this is an action on the deployer, not the deployment which is one layer higher, but I might be missing context here.  

I will try to dig for it somewhere, but highlighting it in this KEP would help other layered operators understand what's required to support a feature IMO.

dhellmann replied:
Yes, all of this implies that the operators need to be aware of the fact that they may be asked to run in a non-HA configuration and that they may need to do something differently with their operand in that case. OLM's responsibility would be to convey the setting to the operator _somehow_ (the proposal is via an environment variable on the deployment of the operator that OLM is managing). After that, it would be up to the author of the operator to make any required changes.

markmc commented:
Seems like pretty clear consensus to just reuse Infrastructure. In which case, it would be good to expand its description beyond just `Infrastructure holds cluster-wide information about Infrastructure` to help people mentally model what belongs there.

Other than that, seems like a sensible approach to the `highAvailabilityMode` challenge ... +1

dhellmann commented:
I have updated the doc to describe adding 1 new field to the existing Infrastructure API. I also renamed the file as part of re-titling the enhancement, so some of the existing comment threads are outdated.

dhellmann commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,347 @@
+Clusters deployed using 4.7 or earlier will not have the new field in
+the `infrastructure` API so the value in the struct will be
+empty. Operators that need the high-availability mode should treat an
+empty string as `Full` high-availability mode.
```
Is this version skew strategy sufficient?

dhellmann commented:
@ravidbro the changes in the latest draft of this enhancement will have an impact on the WIP work you're doing in https://github.com/openshift/api/pull/816

dhellmann commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+values are not set, a replacement for unsupported APIs, and a way to
+avoid adding new operator-specific APIs.
+
+### Operator Lifecycle Manager Operators
```
@mrunalp @ecordell does this wording match what you expect based on our conversation?

dhellmann replied:
Pinging @mrunalp and @ecordell for confirmation about the wording here.

mrunalp replied:
It looks good to me per the discussion we had.

dhellmann commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+to implement but which would nonetheless be easy to overlook when
+implementing a feature.
+
+### Configure OLM-managed operators without the Infrastructure API
```
@mrunalp @ecordell please look at this section, too

romfreiman replied:
I think @ecordell mentioned that there is already an api that the operators can access. Maybe OLM can query Infrastructure and fill it?

dhellmann replied:
That's not what we discussed today. See option 2 on line 354.

soltysh commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+is expanded to include a new field to describe the high-availability
+mode desired for cluster operators running on the cluster.
+
+The `highAvailabilityMode` field will use an enum to express the
```
I'm not quite sure I understand how splitting the profile into two fields, one which seems to be bool (it's HA or not, there's no other option) is unnecessarily introduced as enum, unless I'm missing some more argument what different HA mode we're planning, we'll help here. 
Either way, profile or capabilities, operators will need to know how and what to do with their operands based on either of the fields. Why than overcomplicate this API by adding those extra layers?

markmc replied:
Only one field is proposed - the HA mode field

It's an enum because a previous reviewer (Trevor, Clayton?) recommended never using a bool field because it can't be expanded in the future

Maybe you've interpreted this as meaning there will also be a "cluster profile" field for operators to react to? (Just guessing). While cluster profiles are mentioned above, the intention AIUI is that the HA mode field is the only field that operators should react to.

soltysh replied:
Fair, I think I misread the capabilities from alternatives :+1:

soltysh replied:
Although, honestly this approach is going towards profiles more than just `haMode`. I'm pretty sure it's a matter of time when this is how this will be interpreted. Why not going then with a profile, deploymentMode, or something that is describing the what the cluster will look like, rather than focusing only on HA.

dhellmann replied:
We're not using a profile-name-based API because we don't want to add a bunch of profiles. Defining and maintaining them is a lot of work.

Adding more focused APIs means that only the operators that need to respond to those APIs have to be updated, and (depending on what the API does) they don't necessarily need to be updated all at the same time.

cgwalters replied:
Personally I think we're bikeshedding this particular field a bit too much - yes, we'll need to maintain it ~forever and maybe later we could end up wanting something more sophisticated but... in the end, IMO by *far* the largest maintenance burden isn't going to be these API fields, it's going to be all the work in the operators to honor them and the CI jobs to test different variants, etc.

Whether it's a bool or a string or a list of profiles/flags is really just a small tip of the iceberg - IMO let's just be sure it's not *wrong* of course and then move onto thinking about all those harder problems :wink:

smarterclayton replied:
We don't allow booleans generally in long term apis (and we never have), see Kube API conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types.  No one should add booleans to spec fields without the belief we would never have another mode.  And I don't think even from the discussion so far that we've established that.

If we want to be precise, we're trying to target a specific behavior "single node" which is NOT the same as "single master, many workers" (this configuration could very well want ingress controllers to be highly available).  So we should be explicit about what the user gets, and define a policy that is scoped to what we want, and if we come up with use cases in the future.  I don't think `haMode` is going to be the best name here - if we're being precise (and we should be!), we're describing the preferred cluster topology for the control plane and nothing more, so I would suggest exploring naming towards things like:

`controlPlaneTopology` (topology is a bit more specific than policy, but there may be other words) with values 

* `Singleton` or `SingleNode` or `Minimal` (no strong preference, but each of these is describing the desired topology)
* `HighlyAvailable` (the current setting, which attempts to be highly available out of the box)

We already support additional nodes on HighlyAvailable clusters, and scheduling policy is orthogonal to the setting (these are good things).  If we in the future decide to allow node addition to single node control plane, we don't have to change the constant because we're describing the control plane topology.  If we decide that we need to embody in the product a built-in failover mode (vs having it layered on top), then a new enum value would be `ActivePassive` or `MinimalFailover` or `SingleNodeFailover`.  

It's always better to describe using the enum the thing you expect to see at the outcome as the user would see it.  We know other flags will control other aspects.  That's partially why I don't like `None` - it doesn't describe the current state, whereas a more specific enum value and a more policy focused field name is part of our api approach.

dhellmann replied:
`ActivePassive`, `MinimalFailover`, and `SingleNodeFailover` all sound more like HA modes than topologies to me. We could call the field `haPolicy` I guess?

What sorts of changes would we anticipate having to make to our existing operators to support those new values?

If we want to focus on the topology, would it be better to reflect the `ControlPlane.Replicas` value from the install-config as a number instead of using an enum? And then deal with HA behaviors with an additional field later? I could, for example, see use in a configuration with 3 etcd replicas inside the cluster and HA for routing or API services handled outside of the cluster.

cgwalters replied:
I like `controlPlaneTopology: SingleNode` personally.

>  If we in the future decide to allow node addition to single node control plane,

That...hurts my head.

derekwaynecarr replied:
knowing the topology would provide a nice hint for our web console to not do things that have been problematic when there are no control plane hosts visible in cluster.

derekwaynecarr replied:
So maybe the following:

ControlPlaneTopology: Standalone | External
ControlPlaneAvailability: HA | Minimal

above would be used by things like etcd-operator, auth-operator, stuff that expect to never run on role=worker.


InfrastructureAvailability: HA | Minimal

covers components that target role=worker or infra

This would let us support an IBM Cloud Satellite cluster with 1 worker whose control plane is externalized by not telling the router to expect HA, or the monitoring stack should reduce its footprint AND support the standalone topology needs for single-node or ha.

smarterclayton replied:
oh, external.  Hrm.  If it's external, you don't get to say what the availability is inside the cluster? CPA seems irrelevant for external.

smarterclayton replied:
I'm not against InfrastructureAvailability but I hate infrastructure as a word generally, and availability is wierd.  Hrm...

```
controlPlaneTopology: SingleNode|HighlyAvailable|External
infrastructureTopology: Minimal|HighlyAvailable
```

dhellmann replied:
Does that mean we want to combine topology and availability into 1 field after all? Is there an "external" mode for the infrastructure services, too?

smarterclayton replied:
(derek and i couldn't think of a better word than infrastructure in slack either, and controlPlane and infrastructure are words we use elsewhere so it would be consistent to use them here).  Also in the future we could have external infrastructure but it's unlikely ALL of it would be external

dhellmann replied:
OK, so 2 enum fields, and for now only the control plane supports external.

Do we need to add the external value as part of the first implementation, or can we leave that for a follow-up enhancement for the team working on the external control plane?

ironcladlou replied:
What would External/External mean?

What's the distinction between control plane and infrastructure conceptually when using SingleNode/SingleNode?

smarterclayton replied:
We know there is a possibility (unlikely, let's say 10-40%) that we may in the future want to support single node control planes but multiple infra.  Likewise, we have a use case for External|SingleNode.  We will have a use case for External|HighlyAvailable.  We probably won't have a biz use case for HighlyAvailable|SingleNode, but occasionally people set up small test environments (compact has mostly removed the need for this and i suspect hypershift will remove the rest of the need for it in the future, i.e. External|SingleNode).  External|External might be used (hypothetically) by third party operators that have an integration with Prometheus to directly access data in alertmanager via pod exec (i.e. using infrastructure access rather than api access, which we would generally frown on anyway, but eh).

External would be a followup whenever that relevant feature lands.  You may want to describe "potential future values of this flag" here so people understand the space you are creating and we can refer back to it in later enhancements.  You might mention here that "a future revision may desire to describe an active passive cluster, in which case controlPlaneTopology might acquire a new enumeration `ActivePassive` if such functionality is required"

marun replied:
@smarterclayton Where can I learn more about 'compact'?

smarterclayton replied:
https://github.com/openshift/enhancements/blob/93efdd811027688182575aec7baaa476279705c3/enhancements/compact-clusters.md

Which after this merges will need to be updated to reference what we agree on here

dhellmann replied:
The latest updates to the text have broken the link to this thread, so I am going to resolve it. We can continue the discussion with a new thread that GitHub can show in the right place on the text.

sdodson commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+values are not set, a replacement for unsupported APIs, and a way to
+avoid adding new operator-specific APIs.
+
+### Operator Lifecycle Manager Operators
```
This section seems at odds with https://github.com/openshift/enhancements/pull/560/files#diff-42a0d68eba341abaea58087e60f145ff6b5a01fba2fa2562c3038b831044a82bR290 where it sounds like we're saying OLM will need to set environment variables for Operators because OLM Operators may not necessarily be OpenShift specific. Which is accurate or am I misunderstanding this section?

dhellmann replied:
You're right. This section is correct, the other enhancement needs to be updated to reflect the updated approach.

dhellmann replied:
I've updated the other enhancement.

hexfusion reviewed (commented):
My concern, in general, is that it seems this initial implementation of the API is defining the desired state of `highAvailabilityMode` but as a static `status`. Operators such as `cluster-etcd-operator` must still `ensure` HA is valid during install. (does 3 nodes actually exist) anyways. 

It must be clear that this value is not reflective of infrastructure HA status at all, yet just a desired state configuration. Since the value is not really backed with any guarantee the desired HA config is seemingly only valid during install which makes me wonder why it is part of the API.

smarterclayton commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+window of time between when the API itself is updated and when the
+operators using the API are updated. To avoid issues during that time,
+the controller that manages the infrastructure API will wait for all
+`ClusterOperator` resources to have the same version and not report
```
Why would it check this?  It should just check the cluster version object, which already summarizes that outcome.

dhellmann replied:
Tunnel vision on the part of the author. I'll update this section.

smarterclayton commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+to exist before deploying its operand. If it counted nodes, it could
+not tell the difference between a cluster starting to form with 1
+node, then more, or a cluster that is only ever going to include 1
+node.
```
I think the key question here is whether the amount of masters is orthogonal to the desired state.  Given that we don't want a cluster of size 3 accidentally reduced to 1 to flip into non-HA mode, we should not use "number of masters" to indicate the setting.  We could maybe come up with behaviors that say "revert to being quorum of 1 if you only have one node", but we don't *want* those in failure scenarios.

dhellmann replied:
See also the discussion of why the input replica count does not cover all cases in the alternative described on lines 288-306.

hexfusion replied:
>  Given that we don't want a cluster of size 3 accidentally reduced to 1 to flip into non-HA mode

If we can not describe what HA is in the cluster topology during runtime (masters/machines/X) how do we know if the cluster is actually HA? Today operators need to count master nodes. I would expect if the HA mode could not be enforced some infra controller would be degraded communicating this to the admin/cluster. Otherwise, operators will need to use node counts to go degraded themselves.

I am operator HA Mode is `None` master node count is 2 who is degraded? It's these corner cases that make me nervous.

hexfusion replied:
Should etcd operator own degraded status in this case?

marun replied:
Does an operator have a way of discovering how many nodes a control plane is intended to have?

dhellmann replied:
> Does an operator have a way of discovering how many nodes a control plane is intended to have?

There isn't anything today or in this proposal that would allow that. Do operators need to know the actual number?

hexfusion replied:
> Do operators need to know the actual number?
Some do specifically etcd-operator is concerned with the definition of HA. Our definition of HA is a guarantee that the cluster can provide fault tolerance, which allows for rolling reboots of nodes without quorum loss. The guarantee is directly related to the number of control-plane nodes today. less than 3 HA `None`.

ironcladlou replied:
> > Do operators need to know the actual number?
> 
> Some do specifically etcd-operator is concerned with the definition of HA. Our definition of HA is a guarantee that the cluster can provide fault tolerance, which allows for rolling reboots of nodes without quorum loss. The guarantee is directly related to the number of control-plane nodes today. less than 3 HA `None`.

Is the actual variable driving etcd HA properties the number of control plane nodes, or the number of etcd members, which happen to be (in the stock OCP implementation) coupled to control plane nodes?

Tangentially, other implementations where HA etcd can be achieved without such coupling: containerized etcd, externalized etcd.

ironcladlou replied:
Would it be useful to have a separate discussion elsewhere about the perceived requirement that the etcd-operator be aware of control plane replica counts? The assumption seems to be that replica awareness is essential, which could be worth challenging with fresh eyes just in case it's really not

dhellmann replied:
Yes, I think the discussion of operators needing to know the actual expected replica count should be separate from this change. For the purposes of this enhancement, we only need to be able to differentiate between 1 and 3. Four shalt thou not count, neither count thou two, excepting that thou then proceed to three.

hexfusion replied:
> Would it be useful to have a separate discussion elsewhere about the perceived requirement that the etcd-operator be aware of control plane replica counts? The assumption seems to be that replica awareness is essential, which could be worth challenging with fresh eyes just in case it's really not

I look forward to this discussion.

dhellmann commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+
+### Manage upgrades by tracking settings in each consuming operator
+
+When we make the API mutable we will need to protect against older
```
I'm not sure what you mean by "misclassification"?

hexfusion replied:
If we scale from 3 nodes to 1 or vice versa the cluster will try to enforce an invalid HA mode (misclassified) for the cluster topology. Sometime soon we will need to support 1 node and 3 node clusters. Will it not be supported for a single node UPI cluster to scale up to 3? Because the HA mode is static operators will try to enforce HA mode that is not enforceable given the node count.

failure case: single node cluster scales to 3, then upgrades. But because it was created as a single node it does not have a quorum-guard. Cluster loses quorum during upgrade. Who owns that bug?

It seems like this infra controller should go degraded saying HA mode not met. But in this case customer doesn't want `None` they want `Full`. So it seems that until such time that HA Mode is Mutable or a controller exists that will update status based on topology scaling from 1 to 3 nodes is not supported?

dhellmann replied:
OK, I think I see what you're saying.

If a single-node deployment scales the _control plane_ up then I would expect the relevant operators to scale their operands, including deploying quorum guard for etcd. That implies this API needs to be tied to the one that scales the control plane, somehow. I don't think we have concrete plans to handle this case, but we can make a note that it's going to take significant thought to work out implications like this.

We do have plans to deal with the case where a single-node deployment gains some worker nodes. In that configuration, I wouldn't expect a change in the control plane operands because those worker nodes shouldn't run the control plane components.

hexfusion replied:
> If a single-node deployment scales the control plane up then I would expect the relevant operators to scale their operands, including deploying quorum guard for etcd. 

thanks for the comments. this is my point I guess. The value of the API is arbitrary post-install so the operator can not trust its value to reconcile against as we must still count nodes to verify our actions. These actions "quorum-guard deployment replica 3"  would go against the expectations of HA Mode  `None`.

I understand more work is required but if we can consider this case moving forward I am happy. thanks

dhellmann replied:
I've added this to the risks section above around line 158. Please let me know if I captured your concerns accurately there.

crawford commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,360 @@
+potential parameters to later iterations, we give ourselves more time
+to consider the design. We could also use operator-specific APIs, but
+that is likely to require more work to ensure they are all set
+properly (see the alternatives list below).
```
Is that invalid though? I might want to update my cluster really quickly and would be willing to take downtime to make it happen. Disabling HA would be a great way to do that.

ironcladlou replied:
If you mean invalid in the sense of schema, hopefully literally incoherent values (or combinations of values) could be guarded by validation.

In the sense @crawford seems to be discussing, I agree and think maybe the words "risk" and "invalid" are unnecessarily loaded framings. Each "mode" implies (hopefully explicitly) a set of properties the cluster will exhibit as a result, each with their own potential costs/benefits in the context of a use case.

So one isn't intrinsically more valid than the other in a vacuum: they are merely different and should be clearly described such that they can be understood in comparison to one another by the end user (who is the ultimately the one positioned to evaluate an option's validity for the use case).

Just some thoughts...

dhellmann replied:
I added this concern to the risks section.

marun reviewed (commented):
The proposed API-only approach works great for operators that should be running in a given configuration, but not so great for operators that should not run. Maybe a hybrid approach is suggested?

- Add an API for communicating 'control plane topology' as proposed in this enhancement
  - Only operators that need to respect variance in the API state would need to be updated
- Also add a profile mechanism for manifest exclusion (similar to what ROKS uses)
  - Only operator manifests that should be excluded for a given profile would need to be annotated
  - Annotation-based manifest exclusion is likely cheaper than having to disable operands in operator code
  - Unwanted operators would incur zero runtime cost vs always having to run the operator

marun commented:
@derekwaynecarr I understand you may have strong opinions on supporting an opt-out mechanism for manifests. 

 - Thoughts on what I've suggested in a [previous comment](https://github.com/openshift/enhancements/pull/555#pullrequestreview-564049266)?
- In the absence of a mechanism supporting manifest exclusion, how would a cluster minimize control plane resource consumption?

dhellmann commented:
> @derekwaynecarr I understand you may have strong opinions on supporting an opt-out mechanism for manifests.
> 
> * Thoughts on what I've suggested in a [previous comment](https://github.com/openshift/enhancements/pull/555#pullrequestreview-564049266)?
> * In the absence of a mechanism supporting manifest exclusion, how would a cluster minimize control plane resource consumption?

@marun We think we can build useful single-node deployments with all of the operators present in the system. As you point out, the result is not likely to meet the resource constraints of some (a lot? most?) users, and we plan to address that issue separately in another enhancement. We may use a profile (either in its current form or using exclusion rules as you suggest), or we may add more APIs to disable individual or sets of operands (there's already a way to disable the console, for example).

marun commented:
> > @derekwaynecarr I understand you may have strong opinions on supporting an opt-out mechanism for manifests.
> > 
> > * Thoughts on what I've suggested in a [previous comment](https://github.com/openshift/enhancements/pull/555#pullrequestreview-564049266)?
> > * In the absence of a mechanism supporting manifest exclusion, how would a cluster minimize control plane resource consumption?
> 
> @marun We think we can build useful single-node deployments with all of the operators present in the system. As you point out, the result is not likely to meet the resource constraints of some (a lot? most?) users, and we plan to address that issue separately in another enhancement. We may use a profile (either in its current form or using exclusion rules as you suggest), or we may add more APIs to disable individual or sets of operands (there's already a way to disable the console, for example).

I suggested annotation-based manifest exclusion because it seemed simplest, but have been involved in discussions today indicating that it should always be possible to enable a feature like etcd quorum guard or kube storage migration post-install. That wouldn't be possible if their manifests were simply excluded. That may indicate that an API-only based approach as this enhancement proposes is the way forward, and that any requirement to minimize resource utilization will need to be addressed without manifest exclusion (e.g. by re-homing  optional operands to mandatory operators). In the interests of minimizing unnecessary work, it might be worth finding consensus on this issue in advance of updating optional operators to disable their operands.

dhellmann commented:
I have updated the text of this enhancement to use the 2 enum approach suggested by @smarterclayton and @derekwaynecarr. If you provided an earlier review, please take another look at the latest draft.

I would especially welcome feedback about any inconsistencies that I may have introduced in the latest edits. Thanks!

romfreiman commented:
Thanks you everybody for the comments and the review. It would be great if we'll be able to merge the enhancement till Tuesday-Wednesday and then we'll have couple of days to update all the single-node related PRs accordingly and have them merged once the masters are open, and start running the CIs.

romfreiman commented on enhancements/cluster-high-availability-mode-api.md:
```diff
@@ -0,0 +1,401 @@
+cluster through "topology" values for the control plane and
+infrastructure, separately.
+
+The `controlPlaneTopology` field will express the expectations for
```
how does the installer expected to set those values:
masters=1,workers=0: controlPlaneTopology=SingleNode,infrastructureTopology=None?
masters=1,workers=1: controlPlaneTopology=SingleNode,infrastructureTopology=SingleNode
masters=1,workers>=2: controlPlaneTopology=SingleNode,infrastructureTopology=HighlyAvailable

and if the default is HighlyAvailable, how will we upgrade master=3,worker=1 cluster? (if there are such)

dhellmann replied:
Today a cluster with 3 control plane nodes and 1 worker nodes is not supported in the installer. It fails to deploy because ingress can't create enough replicas of its operand.

dhellmann replied:
I've added some prose explaining the rules for setting the 2 topology fields on lines 97-107 of the latest draft. Please review.

markmc commented:
> I have updated the text of this enhancement to use the 2 enum approach suggested by @smarterclayton and @derekwaynecarr. If you provided an earlier review, please take another look at the latest draft.
> 
> I would especially welcome feedback about any inconsistencies that I may have introduced in the latest edits. Thanks!

I think you've done a nice job capturing the feedback on control-plane versus infrastructure topology fields, so IMO we should be good with merging at this stage. Those nuances will be looked over carefully again in openshift/api#827 and openshift/installer#4497 in case something subtle was overlooked.

/approved

@derekwaynecarr please lgtm if you're good with the latest

markmc commented:
/approve

derekwaynecarr commented:
@dhellmann I had debated if `SingleNode` should be `SingleReplica` or just `Single` , but as it turns out there isn't a great term for the opposite of HA that fits everywhere.  The only nit with `Node` suffix is that we are describing how to schedule pods at the end of the day.  I don't think its worth bike-shedding over, but just remarking here if you wanted to update later.

/approve
/lgtm

dhellmann commented:
> @dhellmann I had debated if `SingleNode` should be `SingleReplica` or just `Single` , but as it turns out there isn't a great term for the opposite of HA that fits everywhere. The only nit with `Node` suffix is that we are describing how to schedule pods at the end of the day. I don't think its worth bike-shedding over, but just remarking here if you wanted to update later.
> 
> /approve
> /lgtm

I could go along with changing it to `SingleReplica`. I'll send another PR.
